plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/credentials/LoginUtils.kt

```diff
@@ -0,0 +1,207 @@
+import software.amazon.awssdk.services.ssooidc.model.SsoOidcException
+import software.amazon.awssdk.services.sts.StsClient
+import software.aws.toolkits.core.region.AwsRegion
+import software.aws.toolkits.core.utils.error
+import software.aws.toolkits.core.utils.tryOrNull
+import software.aws.toolkits.jetbrains.core.AwsClientManager
+import software.aws.toolkits.jetbrains.core.credentials.profiles.SsoSessionConstants
+import software.aws.toolkits.jetbrains.core.credentials.sono.SONO_REGION
+import software.aws.toolkits.jetbrains.core.credentials.sono.SONO_URL
+import software.aws.toolkits.jetbrains.core.credentials.sso.bearer.InteractiveBearerTokenProvider
+import software.aws.toolkits.jetbrains.utils.runUnderProgressIfNeeded
+import software.aws.toolkits.resources.message
+import software.aws.toolkits.telemetry.CredentialSourceId
+import java.io.IOException
+
+private val LOG = LoggerFactory.getLogger("LoginUtils")
+
+sealed class Login {
+    abstract val id: CredentialSourceId
+
+    data class BuilderId(
+        val scopes: List<String>,
+        val onPendingToken: () -> Unit
+    ) : Login() {
+        override val id: CredentialSourceId = CredentialSourceId.AwsId
+
+        fun loginBuilderId(project: Project): Boolean {
+            onPendingToken()
+
+            loginSso(project, SONO_URL, SONO_REGION, scopes)
+            return true
+        }
+    }
+
+    data class IdC(
+        val profileName: String,
+        val startUrl: String,
+        val region: AwsRegion,
+        val scopes: List<String>,
+        val onPendingToken: (InteractiveBearerTokenProvider) -> Unit,
+        val onError: (String) -> Unit
+    ) : Login() {
+        override val id: CredentialSourceId = CredentialSourceId.IamIdentityCenter
+        private val configFilesFacade = DefaultConfigFilesFacade()
+
+        fun loginIdc(project: Project): AwsBearerTokenConnection? {
+            // we have this check here so we blow up early if user has an invalid config file
+            try {
+                configFilesFacade.readSsoSessions()
+            } catch (e: Exception) {
+                println("Failed to read sso sessions file")
+                return null
+            }
+
+            val profile = UserConfigSsoSessionProfile(
+                configSessionName = profileName,
+                ssoRegion = region.id,
+                startUrl = startUrl,
+                scopes = scopes
+            )
+
+            val conn = authAndUpdateConfig(project, profile, configFilesFacade, onPendingToken, onError) ?: return null
+
+            // TODO: delta, make sure we are good to switch immediately
+//            if (!promptForIdcPermissionSet) {
+//                ToolkitConnectionManager.getInstance(project).switchConnection(connection)
+//                close(DialogWrapper.OK_EXIT_CODE)
+//                return
+//            }
+            ToolkitConnectionManager.getInstance(project).switchConnection(conn)
+
+            return conn
```

Comment:
nah, I think the initial intention for not switching connection immediately for `authAndUpdateConfig` is because "promptForIdCPermissionSet". 

So basically if the bearer connection is for `promptForIdCPermissionSet`, it's a magic connection whose purpose is only to fetch sso-sessions so that users can select and use aws explorer.

Reply:
we could keep it as exactly as it was before, but to that end I need to move RoleSelectionPopup and some other stuff into core too